timemachine/cpp/src/kernels/k_nonbonded.cuh

```diff
@@ -169,7 +171,11 @@ void __global__ k_gather(
         return;
     }
 
-    gathered_array[idx * stride + stride_idx] = array[idxs[idx] * stride + stride_idx];
+    // Coords have 3 dimensions, params have 4
+    if (stride_idx < 3) {
+        gathered_coords[idx * 3 + stride_idx] = coords[idxs[idx] * 3 + stride_idx];
+    }
+    gathered_params[idx * stride + stride_idx] = params[idxs[idx] * stride + stride_idx];
```

Comment:
This seems slightly more awkward before (and specializes from a generic function that might be useful elsewhere). Are we sure this is worth a speedup of <1%? (I don't have a super strong opinion here)

Reply:
At this stage we are exchanging generic kernels for performant kernels. Not a massive win here, but see optimizations that make the code less generic become more common going forward. 

While this is not a huge win for our current way of running MD, it will make a bigger difference for the Ixn FF split, where there are 3 of these kernels running (all pairs plus 2 ixn groups). And the recent performance changes are in part to get the ixn split to be closer in performance to the unsplit version.

Reply:
I think this case is find and doesn't really harm readability. For more complicated situations, could it be a viable solution to define more general building blocks as `__device__` functions? For example in this case could we define `k_gather_coords_and_params` in terms of `k_gather` like:

``` cpp
template <typename RealType>
void __device__ k_gather(
    const int N,
    const int stride,
    const unsigned int *__restrict__ idxs,
    const RealType *__restrict__ in,
    RealType *__restrict__ out) {

    int idx = blockIdx.x * blockDim.x + threadIdx.x;
    int stride_idx = blockIdx.y;

    if (idx >= N || stride_idx >= stride) {
        return;
    }

    out[idx * stride + stride_idx] = in[idxs[idx] * stride + stride_idx];
}

template <typename RealType>
void __global__ k_gather_coords_and_params(
    const int N,
    const unsigned int *__restrict__ idxs,
    const RealType *__restrict__ coords,
    const RealType *__restrict__ params,
    RealType *__restrict__ gathered_coords,
    RealType *__restrict__ gathered_params) {

    // Coords have 3 dimensions, params have 4
    k_gather(N, 3, idxs, coords, gathered_coords);
    k_gather(N, 4, idxs, params, gathered_params);
}
```

(Mostly just curious, as this seems like a nice pattern to avoid too much tangling of separate concerns when we get to optimizing more complicated kernels)

Reply:
That would definitely be viable, though would need to investigate what impacts, if any, that has on the number of registers each kernel has.

Will play around

Reply:
Playing around, it seems to be a bit slower to do it with a separated device function, but doesn't impact the number of registers.

Reply:
Still 5-10% slower. It's pretty miniscule, but the generalization gained doesn't seem worth the change.